client/src/document/organisms/sidebar/SidebarFilterer.ts

```diff
@@ -0,0 +1,213 @@
+      delete detail.dataset.open;
+    }
+  }
+
+  private resetHighlighting(link: HTMLAnchorElement) {
+    const nodes = Array.from(link.querySelectorAll<HTMLElement>("span, mark"));
+    const parents = new Set<HTMLElement>();
+    nodes.forEach((node) => {
+      const parent = node.parentElement;
+      node.replaceWith(document.createTextNode(node.textContent ?? ""));
+      if (parent) {
+        parents.add(parent);
+      }
+    });
+    parents.forEach((parent) => parent.normalize());
+  }
+
+  showOnlyMatchingItems(query: string) {
+    this.headings.forEach(this.hideHeading);
+    this.parents.forEach(this.collapseDetail);
+
+    // Show/hide items (+ show parents).
+    const terms = splitQuery(query);
+    let matchCount = 0;
+    this.links.forEach((link) => {
+      this.resetHighlighting(link);
+      const haystack = link.innerText.toLowerCase();
+      const isMatch = terms.every((needle) => haystack.includes(needle));
+
+      const target = link.closest("li") || link;
+      target.style.display = isMatch ? "inherit" : "none";
+
+      if (isMatch) {
+        matchCount++;
+        this.highlightMatches(link, terms);
+        this.showHeading(target);
+        this.expandParents(target);
+      }
+    });
+
+    return matchCount;
+  }
+
+  private hideHeading(heading: HTMLElement) {
+    const container = heading.closest("li") ?? heading;
+    container.style.display = "none";
+  }
+
+  private collapseDetail(el: HTMLDetailsElement) {
+    el.style.display = "none";
+    el.dataset.open = el.dataset.open ?? String(el.open);
+    el.open = false;
+  }
+
+  private highlightMatches(el: HTMLElement, terms: string[]) {
+    const nodes = this.getTextNodes(el);
+
+    nodes.forEach((node) => {
+      const haystack = node.textContent?.toLowerCase();
+      if (!haystack) {
+        return;
+      }
+
+      const ranges = new Map<number, number>();
+      terms.forEach((needle) => {
+        const index = haystack.indexOf(needle);
+        if (index !== -1) {
+          ranges.set(index, index + needle.length);
+        }
+      });
+      const sortedRanges = Array.from(ranges.entries()).sort(
+        ([x1, y1], [x2, y2]) => x1 - x2 || y1 - y2
+      );
+
+      const span = this.replaceChildNode(node, "span");
+      span.className = "highlight-container";
+
+      let rest = span.childNodes[0] as Node & Text;
+      let cursor = 0;
+
+      for (const [rangeBegin, rangeEnd] of sortedRanges) {
+        if (rangeBegin < cursor) {
+          // Just ignore conflicting range.
+          continue;
+        }
+
+        // Split.
+        const match = rest.splitText(rangeBegin - cursor);
+        const newRest = match.splitText(rangeEnd - rangeBegin);
+
+        // Convert text nodes to HTML elements.
+        this.replaceChildNode(rest, "span");
+        this.replaceChildNode(match, "mark");
+
+        rest = newRest as Element & Text;
+        cursor = rangeEnd;
+      }
+
+      this.replaceChildNode(rest, "span");
```

Comment:
Indeed no longer needed, since I wrap the whole text in a span to begin with. I didn't do that at first, and `test <mark>example</mark> and so on` was swallowing the spaces around the mark element.

Fixed in 623bdabed.